File: npm/src/me-uri.ts
export type MeCanonicalSelectorKind = "fanout" | "current" | "surface" | "claim";

export interface MeCanonicalNamespace {
  handle: string;
  space: string;
  value: string;
}

export interface MeCanonicalSelector {
  kind: MeCanonicalSelectorKind;
  raw: string;
  value: string;
  shorthand: boolean;
}

export interface MeCanonicalPath {
  value: string;
  segments: string[];
}

export interface MeCanonicalUri {
  scheme: "me";
  raw: string;
  href: string;
  namespace: string;
  handle: string;
  space: string;
  selector: MeCanonicalSelector | null;
  path: string | null;
  segments: string[];
}

export interface ParseCanonicalMeUriOptions {
  knownSpaces?: readonly string[];
}

export interface CanonicalizeHumanIdentityOptions {
  knownSpaces?: readonly string[];
}

export interface FormatCanonicalMeUriInput {
  handle: string;
  space: string;
  selector?: string | MeCanonicalSelector | null;
  path?: string | null;
}

export type MeDnsProjectionFailureReason =
  | "INVALID_HOST"
  | "TRANSPORT_ONLY_HOST"
  | "UNKNOWN_SPACE"
  | "NOT_CANONICAL_NAMESPACE";

export type MeDnsProjectionResult =
  | {
      ok: true;
      kind: "space";
      rawHost: string;
      host: string;
      matchedSpace: string;
      prefixLabels: [];
      space: string;
    }
  | {
      ok: true;
      kind: "namespace";
      rawHost: string;
      host: string;
      matchedSpace: string;
      prefixLabels: [string];
      handle: string;
      space: string;
      namespace: string;
      uri: string;
    }
  | {
      ok: false;
      kind: "invalid";
      rawHost: string;
      host: string;
      matchedSpace: string | null;
      prefixLabels: string[];
      reason: MeDnsProjectionFailureReason;
    };

export interface MeHumanIdentity {
  raw: string;
  alias: string;
  handle: string;
  space: string;
  namespace: string;
  uri: string;
}

const HANDLE_RE = /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/;
const LABEL_RE = HANDLE_RE;
const SURFACE_NAME_RE = /^[A-Za-z0-9_-]+$/;
const CLAIM_TOKEN_RE = SURFACE_NAME_RE;
const PATH_SEGMENT_RE = /^[A-Za-z0-9_-]+$/;

function assertNonEmpty(rawValue: unknown, label: string): string {
  const value = String(rawValue ?? "").trim();
  if (!value) throw new Error(`${label} is required.`);
  return value;
}

function assertPattern(value: string, pattern: RegExp, label: string, rawValue?: string): string {
  if (!pattern.test(value)) {
    throw new Error(`Invalid ${label}: ${rawValue ?? value}`);
  }
  return value;
}

function normalizeHostishValue(rawValue: unknown): string {
  const raw = String(rawValue ?? "").trim();
  if (!raw) return "";

  try {
    if (/^[a-z]+:\/\//i.test(raw)) {
      return new URL(raw).hostname.trim().toLowerCase();
    }
  } catch {
    // Fall through to manual normalization.
  }

  return raw
    .replace(/^[a-z]+:\/\//i, "")
    .replace(/[/?#].*$/g, "")
    .replace(/:\d+$/g, "")
    .replace(/\.+$/g, "")
    .trim()
    .toLowerCase();
}

function normalizeKnownSpaces(knownSpaces: readonly string[] | undefined): string[] {
  if (!knownSpaces?.length) return [];
  const unique = new Set<string>();
  for (const rawSpace of knownSpaces) {
    unique.add(normalizeCanonicalSpace(rawSpace));
  }
  return Array.from(unique).sort((left, right) => {
    const labelDelta = right.split(".").length - left.split(".").length;
    if (labelDelta !== 0) return labelDelta;
    return right.length - left.length;
  });
}

function parseCanonicalNamespace(rawNamespace: unknown): MeCanonicalNamespace {
  const namespace = assertNonEmpty(rawNamespace, "Canonical namespace")
    .replace(/^me:\/\//i, "")
    .trim()
    .toLowerCase();

  const labels = namespace.split(".").map((label) => label.trim()).filter(Boolean);
  if (labels.length < 3) {
    throw new Error(`Invalid canonical namespace "${rawNamespace}": expected handle.space with a dotted space.`);
  }

  const handle = normalizeCanonicalHandle(labels[0]);
  const space = normalizeCanonicalSpace(labels.slice(1).join("."));
  return {
    handle,
    space,
    value: `${handle}.${space}`,
  };
}

function parseCanonicalSelector(rawSelector: string | null): MeCanonicalSelector | null {
  if (rawSelector === null) return null;
  const value = String(rawSelector).trim();

  if (value === "") {
    return {
      kind: "fanout",
      raw: value,
      value,
      shorthand: false,
    };
  }

  if (value === "current") {
    return {
      kind: "current",
      raw: value,
      value,
      shorthand: false,
    };
  }

  if (value.startsWith("claim:")) {
    const token = assertPattern(value.slice("claim:".length), CLAIM_TOKEN_RE, "claim selector token", value);
    return {
      kind: "claim",
      raw: value,
      value: `claim:${token}`,
      shorthand: false,
    };
  }

  const shorthand = !value.startsWith("surface:");
  const surfaceName = assertPattern(
    shorthand ? value : value.slice("surface:".length),
    SURFACE_NAME_RE,
    "surface selector",
    value,
  );
  return {
    kind: "surface",
    raw: value,
    value: `surface:${surfaceName}`,
    shorthand,
  };
}

function parseCanonicalPath(rawPath: string): MeCanonicalPath {
  const trimmed = String(rawPath ?? "").trim();
  if (!trimmed) {
    throw new Error("Canonical me URI path cannot be empty after '/'.");
  }
  if (trimmed.includes("/")) {
    throw new Error(`Invalid canonical path "${rawPath}": use "." for descent, not "/".`);
  }

  const segments = trimmed.split(".").map((segment) => segment.trim());
  if (segments.some((segment) => !segment)) {
    throw new Error(`Invalid canonical path "${rawPath}": empty segments are not allowed.`);
  }
  for (const segment of segments) {
    assertPattern(segment, PATH_SEGMENT_RE, "canonical path segment", rawPath);
  }

  return {
    value: segments.join("."),
    segments,
  };
}

function serializeSelector(selector: string | MeCanonicalSelector | null | undefined): string {
  if (!selector) return "";
  const parsed = typeof selector === "string" ? parseCanonicalSelector(selector) : parseCanonicalSelector(selector.raw);
  if (!parsed) return "";
  return `[${parsed.value}]`;
}

function findLongestKnownSpaceSuffix(host: string, knownSpaces: readonly string[]): string | null {
  for (const knownSpace of normalizeKnownSpaces(knownSpaces)) {
    if (host === knownSpace || host.endsWith(`.${knownSpace}`)) {
      return knownSpace;
    }
  }
  return null;
}

export function normalizeCanonicalHandle(rawHandle: unknown): string {
  return assertPattern(assertNonEmpty(rawHandle, "Handle").toLowerCase(), HANDLE_RE, "canonical handle", String(rawHandle));
}

export function normalizeCanonicalSpace(rawSpace: unknown): string {
  const normalized = normalizeHostishValue(assertNonEmpty(rawSpace, "Space"));
  const labels = normalized.split(".").map((label) => label.trim()).filter(Boolean);
  if (labels.length < 2) {
    throw new Error(`Invalid canonical space "${rawSpace}": expected at least two labels.`);
  }
  for (const label of labels) {
    assertPattern(label, LABEL_RE, "space label", String(rawSpace));
  }
  return labels.join(".");
}

export function formatCanonicalMeUri(input: FormatCanonicalMeUriInput): string {
  const handle = normalizeCanonicalHandle(input.handle);
  const space = normalizeCanonicalSpace(input.space);
  const selector = serializeSelector(input.selector);
  const path = input.path == null || String(input.path).trim() === ""
    ? ""
    : `/${parseCanonicalPath(String(input.path)).value}`;
  return `me://${handle}.${space}${selector}${path}`;
}

export function parseMeUri(rawInput: string): MeCanonicalUri {
  const raw = assertNonEmpty(rawInput, "me:// URI");
  if (!raw.toLowerCase().startsWith("me://")) {
    throw new Error(`Invalid me URI "${rawInput}": expected "me://" scheme.`);
  }

  const withoutScheme = raw.slice(5);
  if (!withoutScheme.trim()) {
    throw new Error(`Invalid me URI "${rawInput}": missing namespace.`);
  }

  const slashIndex = withoutScheme.indexOf("/");
  const head = (slashIndex >= 0 ? withoutScheme.slice(0, slashIndex) : withoutScheme).trim();
  const pathPart = slashIndex >= 0 ? withoutScheme.slice(slashIndex + 1) : "";

  let namespacePart = head;
  let selectorPart: string | null = null;
  const openIndex = head.indexOf("[");
  if (openIndex >= 0) {
    const closeIndex = head.lastIndexOf("]");
    if (closeIndex < openIndex || closeIndex !== head.length - 1) {
      throw new Error(`Invalid me URI "${rawInput}": malformed selector.`);
    }
    namespacePart = head.slice(0, openIndex).trim();
    selectorPart = head.slice(openIndex + 1, closeIndex).trim();
  }

  const namespace = parseCanonicalNamespace(namespacePart);
  const selector = parseCanonicalSelector(selectorPart);
  const path = slashIndex >= 0 ? parseCanonicalPath(pathPart) : null;
  const href = formatCanonicalMeUri({
    handle: namespace.handle,
    space: namespace.space,
    selector,
    path: path?.value ?? null,
  });

  return {
    scheme: "me",
    raw,
    href,
    namespace: namespace.value,
    handle: namespace.handle,
    space: namespace.space,
    selector,
    path: path?.value ?? null,
    segments: path?.segments ?? [],
  };
}

export function tryParseMeUri(rawInput: string): MeCanonicalUri | null {
  try {
    return parseMeUri(rawInput);
  } catch {
    return null;
  }
}

export function parseCanonicalMeUri(
  rawInput: string,
  options: ParseCanonicalMeUriOptions = {},
): MeCanonicalUri {
  const parsed = parseMeUri(rawInput);
  const knownSpaces = normalizeKnownSpaces(options.knownSpaces);
  if (knownSpaces.length > 0 && !knownSpaces.includes(parsed.space)) {
    throw new Error(`Unknown canonical space "${parsed.space}" in "${rawInput}".`);
  }
  return parsed;
}

export function canonicalizeHumanIdentity(
  rawInput: string,
  options: CanonicalizeHumanIdentityOptions = {},
): MeHumanIdentity {
  const raw = assertNonEmpty(rawInput, "Human identity");
  const parts = raw.split("@");
  if (parts.length !== 2) {
    throw new Error(`Invalid human identity "${rawInput}": expected handle@space.`);
  }

  const handle = normalizeCanonicalHandle(parts[0]);
  const space = normalizeCanonicalSpace(parts[1]);
  const uri = formatCanonicalMeUri({ handle, space });
  parseCanonicalMeUri(uri, options);

  return {
    raw,
    alias: `${handle}@${space}`,
    handle,
    space,
    namespace: `${handle}.${space}`,
    uri,
  };
}

export function projectDnsHostToNamespace(
  rawHost: string,
  knownSpaces: readonly string[],
): MeDnsProjectionResult {
  const host = normalizeHostishValue(rawHost);
  if (!host) {
    return {
      ok: false,
      kind: "invalid",
      rawHost: String(rawHost ?? ""),
      host: "",
      matchedSpace: null,
      prefixLabels: [],
      reason: "INVALID_HOST",
    };
  }

  if (host === "localhost" || host.endsWith(".local")) {
    return {
      ok: false,
      kind: "invalid",
      rawHost,
      host,
      matchedSpace: null,
      prefixLabels: [],
      reason: "TRANSPORT_ONLY_HOST",
    };
  }

  const labels = host.split(".").map((label) => label.trim()).filter(Boolean);
  if (labels.length < 2) {
    return {
      ok: false,
      kind: "invalid",
      rawHost,
      host,
      matchedSpace: null,
      prefixLabels: labels,
      reason: "INVALID_HOST",
    };
  }
  for (const label of labels) {
    if (!LABEL_RE.test(label)) {
      return {
        ok: false,
        kind: "invalid",
        rawHost,
        host,
        matchedSpace: null,
        prefixLabels: labels,
        reason: "INVALID_HOST",
      };
    }
  }

  const matchedSpace = findLongestKnownSpaceSuffix(host, knownSpaces);
  if (!matchedSpace) {
    return {
      ok: false,
      kind: "invalid",
      rawHost,
      host,
      matchedSpace: null,
      prefixLabels: labels,
      reason: "UNKNOWN_SPACE",
    };
  }

  const prefix = host === matchedSpace ? "" : host.slice(0, -(matchedSpace.length + 1));
  const prefixLabels = prefix ? prefix.split(".").filter(Boolean) : [];
  if (prefixLabels.length === 0) {
    return {
      ok: true,
      kind: "space",
      rawHost,
      host,
      matchedSpace,
      prefixLabels: [],
      space: matchedSpace,
    };
  }

  if (prefixLabels.length !== 1) {
    return {
      ok: false,
      kind: "invalid",
      rawHost,
      host,
      matchedSpace,
      prefixLabels,
      reason: "NOT_CANONICAL_NAMESPACE",
    };
  }

  const handle = normalizeCanonicalHandle(prefixLabels[0]);
  const namespace = `${handle}.${matchedSpace}`;
  return {
    ok: true,
    kind: "namespace",
    rawHost,
    host,
    matchedSpace,
    prefixLabels: [handle],
    handle,
    space: matchedSpace,
    namespace,
    uri: formatCanonicalMeUri({ handle, space: matchedSpace }),
  };
}

File: Rust/src/me_uri.rs
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MeCanonicalSelectorKind {
    Fanout,
    Current,
    Surface,
    Claim,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MeCanonicalSelector {
    pub kind: MeCanonicalSelectorKind,
    pub raw: String,
    pub value: String,
    pub shorthand: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MeCanonicalUri {
    pub scheme: String,
    pub raw: String,
    pub href: String,
    pub namespace: String,
    pub handle: String,
    pub space: String,
    pub selector: Option<MeCanonicalSelector>,
    pub path: Option<String>,
    pub segments: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MeHumanIdentity {
    pub raw: String,
    pub alias: String,
    pub handle: String,
    pub space: String,
    pub namespace: String,
    pub uri: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MeDnsProjectionFailureReason {
    InvalidHost,
    TransportOnlyHost,
    UnknownSpace,
    NotCanonicalNamespace,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MeDnsProjection {
    Space {
        raw_host: String,
        host: String,
        matched_space: String,
        prefix_labels: Vec<String>,
        space: String,
    },
    Namespace {
        raw_host: String,
        host: String,
        matched_space: String,
        prefix_labels: Vec<String>,
        handle: String,
        space: String,
        namespace: String,
        uri: String,
    },
    Invalid {
        raw_host: String,
        host: String,
        matched_space: Option<String>,
        prefix_labels: Vec<String>,
        reason: MeDnsProjectionFailureReason,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MeUriError {
    Required(&'static str),
    InvalidScheme(String),
    MissingNamespace(String),
    InvalidCanonicalNamespace(String),
    InvalidCanonicalHandle(String),
    InvalidCanonicalSpace(String),
    InvalidSpaceLabel(String),
    InvalidSelector(String),
    InvalidClaimSelector(String),
    InvalidPath(String),
    UnknownCanonicalSpace(String),
    InvalidHumanIdentity(String),
    MalformedSelector(String),
}

impl fmt::Display for MeUriError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Required(label) => write!(f, "{label} is required"),
            Self::InvalidScheme(value) => {
                write!(f, "invalid me URI \"{value}\": expected \"me://\" scheme")
            }
            Self::MissingNamespace(value) => {
                write!(f, "invalid me URI \"{value}\": missing namespace")
            }
            Self::InvalidCanonicalNamespace(value) => write!(
                f,
                "invalid canonical namespace \"{value}\": expected handle.space with a dotted space"
            ),
            Self::InvalidCanonicalHandle(value) => write!(f, "invalid canonical handle: {value}"),
            Self::InvalidCanonicalSpace(value) => write!(
                f,
                "invalid canonical space \"{value}\": expected at least two labels"
            ),
            Self::InvalidSpaceLabel(value) => write!(f, "invalid space label: {value}"),
            Self::InvalidSelector(value) => write!(f, "invalid surface selector: {value}"),
            Self::InvalidClaimSelector(value) => write!(f, "invalid claim selector token: {value}"),
            Self::InvalidPath(value) => write!(f, "invalid canonical path \"{value}\""),
            Self::UnknownCanonicalSpace(value) => write!(f, "unknown canonical space \"{value}\""),
            Self::InvalidHumanIdentity(value) => {
                write!(
                    f,
                    "invalid human identity \"{value}\": expected handle@space"
                )
            }
            Self::MalformedSelector(value) => {
                write!(f, "invalid me URI \"{value}\": malformed selector")
            }
        }
    }
}

impl std::error::Error for MeUriError {}

pub fn normalize_canonical_handle(raw_handle: &str) -> Result<String, MeUriError> {
    let handle = required(raw_handle, "Handle")?.to_ascii_lowercase();
    if !is_dns_label(&handle) {
        return Err(MeUriError::InvalidCanonicalHandle(raw_handle.to_string()));
    }
    Ok(handle)
}

pub fn normalize_canonical_space(raw_space: &str) -> Result<String, MeUriError> {
    let normalized = normalize_hostish_value(required(raw_space, "Space")?);
    let labels = split_labels(&normalized);
    if labels.len() < 2 {
        return Err(MeUriError::InvalidCanonicalSpace(raw_space.to_string()));
    }
    for label in &labels {
        if !is_dns_label(label) {
            return Err(MeUriError::InvalidSpaceLabel(raw_space.to_string()));
        }
    }
    Ok(labels.join("."))
}

pub fn format_canonical_me_uri(
    handle: &str,
    space: &str,
    selector: Option<&str>,
    path: Option<&str>,
) -> Result<String, MeUriError> {
    let handle = normalize_canonical_handle(handle)?;
    let space = normalize_canonical_space(space)?;
    let selector = serialize_selector(selector)?;
    let path = match path {
        Some(raw_path) if !raw_path.trim().is_empty() => {
            format!("/{}", parse_canonical_path(raw_path)?.0)
        }
        _ => String::new(),
    };
    Ok(format!("me://{handle}.{space}{selector}{path}"))
}

pub fn parse_me_uri(raw_input: &str) -> Result<MeCanonicalUri, MeUriError> {
    let raw = required(raw_input, "me:// URI")?.to_string();
    if !raw.to_ascii_lowercase().starts_with("me://") {
        return Err(MeUriError::InvalidScheme(raw_input.to_string()));
    }

    let without_scheme = &raw[5..];
    if without_scheme.trim().is_empty() {
        return Err(MeUriError::MissingNamespace(raw_input.to_string()));
    }

    let slash_index = without_scheme.find('/');
    let head = slash_index
        .map(|index| &without_scheme[..index])
        .unwrap_or(without_scheme)
        .trim();
    let path_part = slash_index
        .map(|index| &without_scheme[index + 1..])
        .unwrap_or("");

    let mut namespace_part = head;
    let mut selector_part = None;
    if let Some(open_index) = head.find('[') {
        let close_index = head.rfind(']');
        if close_index.is_none_or(|index| index < open_index || index != head.len() - 1) {
            return Err(MeUriError::MalformedSelector(raw_input.to_string()));
        }
        let close_index = close_index.expect("checked above");
        namespace_part = head[..open_index].trim();
        selector_part = Some(head[open_index + 1..close_index].trim());
    }

    let namespace = parse_canonical_namespace(namespace_part)?;
    let selector = parse_canonical_selector(selector_part)?;
    let path = if slash_index.is_some() {
        Some(parse_canonical_path(path_part)?)
    } else {
        None
    };
    let href = format_canonical_me_uri(
        &namespace.handle,
        &namespace.space,
        selector.as_ref().map(|selector| selector.raw.as_str()),
        path.as_ref().map(|path| path.0.as_str()),
    )?;

    Ok(MeCanonicalUri {
        scheme: "me".to_string(),
        raw,
        href,
        namespace: namespace.value,
        handle: namespace.handle,
        space: namespace.space,
        selector,
        path: path.as_ref().map(|path| path.0.clone()),
        segments: path.map(|path| path.1).unwrap_or_default(),
    })
}

pub fn try_parse_me_uri(raw_input: &str) -> Option<MeCanonicalUri> {
    parse_me_uri(raw_input).ok()
}

pub fn parse_canonical_me_uri(
    raw_input: &str,
    known_spaces: &[&str],
) -> Result<MeCanonicalUri, MeUriError> {
    let parsed = parse_me_uri(raw_input)?;
    let known_spaces = normalize_known_spaces(known_spaces)?;
    if !known_spaces.is_empty() && !known_spaces.contains(&parsed.space) {
        return Err(MeUriError::UnknownCanonicalSpace(parsed.space));
    }
    Ok(parsed)
}

pub fn canonicalize_human_identity(
    raw_input: &str,
    known_spaces: &[&str],
) -> Result<MeHumanIdentity, MeUriError> {
    let raw = required(raw_input, "Human identity")?.to_string();
    let mut parts = raw.split('@');
    let handle_part = parts.next();
    let space_part = parts.next();
    if handle_part.is_none() || space_part.is_none() || parts.next().is_some() {
        return Err(MeUriError::InvalidHumanIdentity(raw_input.to_string()));
    }

    let handle = normalize_canonical_handle(handle_part.expect("checked above"))?;
    let space = normalize_canonical_space(space_part.expect("checked above"))?;
    let uri = format_canonical_me_uri(&handle, &space, None, None)?;
    parse_canonical_me_uri(&uri, known_spaces)?;
    let namespace = format!("{handle}.{space}");

    Ok(MeHumanIdentity {
        raw,
        alias: format!("{handle}@{space}"),
        handle,
        space,
        namespace,
        uri,
    })
}

pub fn canonicalize_legacy_at_operator(raw_input: &str, known_spaces: &[&str]) -> Option<String> {
    let raw = raw_input.trim();
    if !raw.contains('@') {
        return None;
    }
    canonicalize_human_identity(raw, known_spaces)
        .ok()
        .map(|identity| identity.uri)
}

pub fn project_dns_host_to_namespace(raw_host: &str, known_spaces: &[&str]) -> MeDnsProjection {
    let host = normalize_hostish_value(raw_host);
    if host.is_empty() {
        return MeDnsProjection::Invalid {
            raw_host: raw_host.to_string(),
            host,
            matched_space: None,
            prefix_labels: Vec::new(),
            reason: MeDnsProjectionFailureReason::InvalidHost,
        };
    }

    if host == "localhost" || host.ends_with(".local") {
        return MeDnsProjection::Invalid {
            raw_host: raw_host.to_string(),
            host,
            matched_space: None,
            prefix_labels: Vec::new(),
            reason: MeDnsProjectionFailureReason::TransportOnlyHost,
        };
    }

    let known_spaces = normalize_known_spaces(known_spaces).unwrap_or_default();
    let Some(matched_space) = find_longest_known_space_suffix(&host, &known_spaces) else {
        return MeDnsProjection::Invalid {
            raw_host: raw_host.to_string(),
            host,
            matched_space: None,
            prefix_labels: Vec::new(),
            reason: MeDnsProjectionFailureReason::UnknownSpace,
        };
    };

    if host == matched_space {
        return MeDnsProjection::Space {
            raw_host: raw_host.to_string(),
            host,
            matched_space: matched_space.clone(),
            prefix_labels: Vec::new(),
            space: matched_space,
        };
    }

    let prefix = host
        .strip_suffix(&format!(".{matched_space}"))
        .unwrap_or("")
        .to_string();
    let prefix_labels = split_labels(&prefix);
    if prefix_labels.len() != 1 || !is_dns_label(&prefix_labels[0]) {
        return MeDnsProjection::Invalid {
            raw_host: raw_host.to_string(),
            host,
            matched_space: Some(matched_space),
            prefix_labels,
            reason: MeDnsProjectionFailureReason::NotCanonicalNamespace,
        };
    }

    let handle = prefix_labels[0].clone();
    let namespace = format!("{handle}.{matched_space}");
    let uri = format_canonical_me_uri(&handle, &matched_space, None, None).unwrap_or_else(|_| {
        // The matched space and handle were normalized above; this path is unreachable.
        format!("me://{namespace}")
    });

    MeDnsProjection::Namespace {
        raw_host: raw_host.to_string(),
        host,
        matched_space: matched_space.clone(),
        prefix_labels,
        handle,
        space: matched_space,
        namespace,
        uri,
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct MeCanonicalNamespace {
    handle: String,
    space: String,
    value: String,
}

fn parse_canonical_namespace(raw_namespace: &str) -> Result<MeCanonicalNamespace, MeUriError> {
    let namespace = required(raw_namespace, "Canonical namespace")?
        .trim_start_matches("me://")
        .trim()
        .to_ascii_lowercase();
    let labels = split_labels(&namespace);
    if labels.len() < 3 {
        return Err(MeUriError::InvalidCanonicalNamespace(
            raw_namespace.to_string(),
        ));
    }
    let handle = normalize_canonical_handle(&labels[0])?;
    let space = normalize_canonical_space(&labels[1..].join("."))?;
    let value = format!("{handle}.{space}");
    Ok(MeCanonicalNamespace {
        handle,
        space,
        value,
    })
}

fn parse_canonical_selector(
    raw_selector: Option<&str>,
) -> Result<Option<MeCanonicalSelector>, MeUriError> {
    let Some(raw_selector) = raw_selector else {
        return Ok(None);
    };
    let value = raw_selector.trim();

    if value.is_empty() {
        return Ok(Some(MeCanonicalSelector {
            kind: MeCanonicalSelectorKind::Fanout,
            raw: value.to_string(),
            value: value.to_string(),
            shorthand: false,
        }));
    }

    if value == "current" {
        return Ok(Some(MeCanonicalSelector {
            kind: MeCanonicalSelectorKind::Current,
            raw: value.to_string(),
            value: value.to_string(),
            shorthand: false,
        }));
    }

    if let Some(token) = value.strip_prefix("claim:") {
        if !is_surface_token(token) {
            return Err(MeUriError::InvalidClaimSelector(value.to_string()));
        }
        return Ok(Some(MeCanonicalSelector {
            kind: MeCanonicalSelectorKind::Claim,
            raw: value.to_string(),
            value: format!("claim:{token}"),
            shorthand: false,
        }));
    }

    let shorthand = !value.starts_with("surface:");
    let surface_name = if shorthand {
        value
    } else {
        &value["surface:".len()..]
    };
    if !is_surface_token(surface_name) {
        return Err(MeUriError::InvalidSelector(value.to_string()));
    }

    Ok(Some(MeCanonicalSelector {
        kind: MeCanonicalSelectorKind::Surface,
        raw: value.to_string(),
        value: format!("surface:{surface_name}"),
        shorthand,
    }))
}

fn parse_canonical_path(raw_path: &str) -> Result<(String, Vec<String>), MeUriError> {
    let trimmed = raw_path.trim();
    if trimmed.is_empty() {
        return Err(MeUriError::InvalidPath(raw_path.to_string()));
    }
    if trimmed.contains('/') {
        return Err(MeUriError::InvalidPath(raw_path.to_string()));
    }

    let segments = trimmed
        .split('.')
        .map(str::trim)
        .map(str::to_string)
        .collect::<Vec<_>>();
    if segments.iter().any(|segment| segment.is_empty()) {
        return Err(MeUriError::InvalidPath(raw_path.to_string()));
    }
    if segments.iter().any(|segment| !is_path_segment(segment)) {
        return Err(MeUriError::InvalidPath(raw_path.to_string()));
    }

    Ok((segments.join("."), segments))
}

fn serialize_selector(selector: Option<&str>) -> Result<String, MeUriError> {
    let Some(selector) = selector else {
        return Ok(String::new());
    };
    let Some(parsed) = parse_canonical_selector(Some(selector))? else {
        return Ok(String::new());
    };
    Ok(format!("[{}]", parsed.value))
}

fn normalize_known_spaces(known_spaces: &[&str]) -> Result<Vec<String>, MeUriError> {
    let mut unique = Vec::<String>::new();
    for raw_space in known_spaces {
        let space = normalize_canonical_space(raw_space)?;
        if !unique.contains(&space) {
            unique.push(space);
        }
    }
    unique.sort_by(|left, right| {
        right
            .split('.')
            .count()
            .cmp(&left.split('.').count())
            .then_with(|| right.len().cmp(&left.len()))
    });
    Ok(unique)
}

fn find_longest_known_space_suffix(host: &str, known_spaces: &[String]) -> Option<String> {
    for known_space in known_spaces {
        if host == known_space || host.ends_with(&format!(".{known_space}")) {
            return Some(known_space.clone());
        }
    }
    None
}

fn normalize_hostish_value(raw_value: &str) -> String {
    let mut value = raw_value.trim().to_string();
    if value.is_empty() {
        return value;
    }

    if let Some(scheme_index) = value.find("://") {
        value = value[scheme_index + 3..].to_string();
    }
    for delimiter in ['/', '?', '#'] {
        if let Some(index) = value.find(delimiter) {
            value.truncate(index);
        }
    }
    if let Some(colon_index) = value.rfind(':') {
        if value[colon_index + 1..]
            .chars()
            .all(|ch| ch.is_ascii_digit())
        {
            value.truncate(colon_index);
        }
    }

    value.trim_end_matches('.').trim().to_ascii_lowercase()
}

fn required<'a>(value: &'a str, label: &'static str) -> Result<&'a str, MeUriError> {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return Err(MeUriError::Required(label));
    }
    Ok(trimmed)
}

fn split_labels(value: &str) -> Vec<String> {
    value
        .split('.')
        .map(str::trim)
        .filter(|label| !label.is_empty())
        .map(str::to_string)
        .collect()
}

fn is_dns_label(value: &str) -> bool {
    let len = value.len();
    if len == 0 || len > 63 {
        return false;
    }
    let bytes = value.as_bytes();
    if bytes[0] == b'-' || bytes[len - 1] == b'-' {
        return false;
    }
    value
        .chars()
        .all(|ch| ch.is_ascii_lowercase() || ch.is_ascii_digit() || ch == '-')
}

fn is_surface_token(value: &str) -> bool {
    !value.is_empty()
        && value
            .chars()
            .all(|ch| ch.is_ascii_alphanumeric() || ch == '_' || ch == '-')
}

fn is_path_segment(value: &str) -> bool {
    is_surface_token(value)
}
